File: app/src/test/java/com/example/david/lists/data/repository/RepositoryTest.java
package com.example.david.lists.data.repository;

import com.example.david.lists.data.datamodel.Item;
import com.example.david.lists.data.datamodel.UserList;
import com.example.david.lists.data.remote.IRemoteRepositoryContract;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyZeroInteractions;

@RunWith(MockitoJUnitRunner.class)
public class RepositoryTest {

    @InjectMocks
    private Repository repository;

    @Mock
    private IRemoteRepositoryContract.Repository remoteDatabase;

    private final String id = "qwerty";
    private final String newTitle = "New Title";
    private final UserList userList = new UserList("title", 0);
    private final Item item = new Item("title", 0, "qwerty");


    @Test
    public void getAllUserLists_InvokesCorrectRemoteStorageMethodOnce() {
        repository.getAllUserLists();
        verify(remoteDatabase, times(1)).getUserLists();
    }

    @Test
    public void getAllItems_InvokesCorrectRemoteStorageMethodOnce() {
        repository.getItems(id);
        verify(remoteDatabase, times(1)).getItems(id);
    }

    @Test
    public void addUserList_InvokesCorrectRemoteStorageMethodOnce() {
        repository.addUserList(userList);
        verify(remoteDatabase, times(1)).addUserList(userList);
    }

    @Test
    public void addItem_InvokesCorrectRemoteStorageMethodOnce() {
        repository.addItem(item);
        verify(remoteDatabase, times(1)).addItem(item);
    }

    @Test
    public void deleteUserLists_InvokesCorrectRemoteStorageMethodOnce() {
        List<UserList> userLists = Collections.singletonList(userList);
        repository.deleteUserLists(userLists);
        verify(remoteDatabase, times(1)).deleteUserLists(userLists);
    }

    @Test
    public void deleteItems_InvokesCorrectRemoteStorageMethodOnce() {
        List<Item> items = Collections.singletonList(item);
        repository.deleteItems(items);
        verify(remoteDatabase, times(1)).deleteItems(items);
    }

    @Test
    public void renameUserList_InvokesCorrectRemoteStorageMethodOnce() {
        repository.renameUserList(id, newTitle);
        verify(remoteDatabase, times(1)).renameUserList(id, newTitle);
    }

    @Test
    public void renameItem_InvokesCorrectRemoteStorageMethodOnce() {
        repository.renameItem(id, newTitle);
        verify(remoteDatabase, times(1)).renameItem(id, newTitle);
    }

    @Test
    public void updateUserListPosition__InvokesCorrectRemoteStorageMethodOnce() {
        int newPosition = 1;
        int oldPosition = 5;
        repository.updateUserListPosition(userList, oldPosition, newPosition);
        verify(remoteDatabase, times(1)).updateUserListPosition(userList, oldPosition, newPosition);
    }

    @Test
    public void updateItemPosition__InvokesCorrectRemoteStorageMethodOnce() {
        int newPosition = 1;
        int oldPosition = 5;
        repository.updateItemPosition(item, oldPosition, newPosition);
        verify(remoteDatabase, times(1)).updateItemPosition(item, oldPosition, newPosition);
    }


    /**ERROR TESTING**/

    @Test(expected = IllegalArgumentException.class)
    public void getItems_whenIdIsEmpty_ThrowsIllegalArgumentException() {
        repository.getItems("");
    }

    @Test(expected = IllegalArgumentException.class)
    public void getItems_WhenIdIsNull_ThrowsIllegalArgumentException() {
        repository.getItems(null);
    }


    @Test(expected = IllegalArgumentException.class)
    public void addUserList_WhenNull_ThrowsIllegalArgumentException() {
        repository.addUserList(null);
    }

    @Test(expected = IllegalArgumentException.class)
    public void addItem_WhenNull_ThrowsIllegalArgumentException() {
        repository.addItem(null);
    }


    @Test(expected = IllegalArgumentException.class)
    public void deleteUserList_WhenListIsEmpty_ThrowIllegalArgumentException() {
        repository.deleteUserLists(new ArrayList<>());
    }

    @Test(expected = IllegalArgumentException.class)
    public void deleteItems_WhenListIsEmpty_ThrowIllegalArgumentException() {
        repository.deleteItems(new ArrayList<>());
    }


    @Test(expected = IllegalArgumentException.class)
    public void renameUserList_WhenTitleIsEmpty_ThrowsIllegalArgumentException() {
        repository.renameUserList("placeholder", "");
    }

    @Test(expected = IllegalArgumentException.class)
    public void renameUserList_WhenTitleIsNull_ThrowsIllegalArgumentException() {
        repository.renameUserList("placeholder", null);
    }

    @Test(expected = IllegalArgumentException.class)
    public void renameUserList_WhenIdIsEmpty_ThrowsIllegalArgumentException() {
        repository.renameUserList("", "newTitle");
    }

    @Test(expected = IllegalArgumentException.class)
    public void renameUserList_WhenIdIsNull_ThrowsIllegalArgumentException() {
        repository.renameUserList(null, "newTitle");
    }

    @Test(expected = IllegalArgumentException.class)
    public void renameItem_WhenTitleIsEmpty_ThrowsIllegalArgumentException() {
        repository.renameItem("placeholder", "");
    }

    @Test(expected = IllegalArgumentException.class)
    public void renameItem_WhenTitleIsNull_ThrowsIllegalArgumentException() {
        repository.renameItem("placeholder", null);
    }

    @Test(expected = IllegalArgumentException.class)
    public void renameItem_WhenIdIsEmpty_ThrowsIllegalArgumentException() {
        repository.renameItem("", "newTitle");
    }

    @Test(expected = IllegalArgumentException.class)
    public void renameItem_WhenIdIsNull_ThrowsIllegalArgumentException() {
        repository.renameItem(null, "newTitle");
    }


    public void updateUserListPosition_WhenPositionsAreTheSame_Returns() {
        repository.updateUserListPosition(new UserList("placeholder", 0), 5, 5);
        verifyZeroInteractions(remoteDatabase);
    }

    @Test(expected = IllegalArgumentException.class)
    public void updateUserListPosition_WhenPositionsAreNegative_ThrowsIllegalArgumentException() {
        repository.updateUserListPosition(new UserList("placeholder", 0), -1, -10);
    }

    @Test(expected = IllegalArgumentException.class)
    public void updateUserListPosition_WhenFirstPositionIsNegative_ThrowsIllegalArgumentException() {
        repository.updateUserListPosition(new UserList("placeholder", 0), -1, 10);
    }

    @Test(expected = IllegalArgumentException.class)
    public void updateUserListPosition_WhenSecondPositionIsNegative_ThrowsIllegalArgumentException() {
        repository.updateUserListPosition(new UserList("placeholder", 0), 1, -1);
    }

    @Test(expected = IllegalArgumentException.class)
    public void updateUserListPosition_WhenNullObject_ThrowsIllegalArgumentException() {
        repository.updateUserListPosition(null, 1, 10);
    }


    public void updateItemPosition_WhenPositionsAreTheSame_Returns() {
        repository.updateItemPosition(new Item("placeholder", 0, "qwerty"), 5, 5);
        verifyZeroInteractions(remoteDatabase);
    }

    @Test(expected = IllegalArgumentException.class)
    public void updateItemPosition_WhenPositionsAreNegative_ThrowsIllegalArgumentException() {
        repository.updateItemPosition(new Item("placeholder", 0, "qwerty"), -1, -10);
    }

    @Test(expected = IllegalArgumentException.class)
    public void updateItemPosition_WhenFirstPositionIsNegative_ThrowsIllegalArgumentException() {
        repository.updateItemPosition(new Item("placeholder", 0, "qwerty"), -1, 1);
    }

    @Test(expected = IllegalArgumentException.class)
    public void updateItemPosition_WhenSecondPositionIsNegative_ThrowsIllegalArgumentException() {
        repository.updateItemPosition(new Item("placeholder", 0, "qwerty"), 1, -1);
    }

    @Test(expected = IllegalArgumentException.class)
    public void updateItemPosition_WhenNullObject_ThrowsIllegalArgumentException() {
        repository.updateItemPosition(null, 1, 10);
    }
}

File: app/src/test/java/com/example/david/lists/data/repository/RepositoryTest.kt
package com.example.david.lists.data.repository

import com.example.david.lists.data.datamodel.Item
import com.example.david.lists.data.datamodel.UserList
import com.example.david.lists.data.remote.IRemoteRepositoryContract
import io.mockk.clearAllMocks
import io.mockk.mockk
import io.mockk.verify
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows

class RepositoryTest {

    private val remoteRepo = mockk<IRemoteRepositoryContract.Repository>(relaxUnitFun = true)

    private val repo = Repository(remoteRepo)

    private val id = "qwerty"
    private val title = "title"
    private val position = 0
    private val userListId = "user_list_id"

    private val userList = UserList(title, position)
    private val item = Item(title, position, userListId)


    @BeforeEach
    fun init() {
        clearAllMocks()
    }


    @Nested
    inner class Delete {
        /**
         * Empty List argument - Exception thrown
         */
        @Test
        fun `Delete UserLists - empty list`() {
            assertThrows<IllegalArgumentException> {
                repo.deleteUserLists(emptyList())
            }
        }


        /**
         * Empty List argument - Exception thrown
         */
        @Test
        fun `Delete Items - EmptyList`() {
            assertThrows<IllegalArgumentException> {
                repo.deleteItems(emptyList())
            }
        }
    }


    @Nested
    inner class Rename {
        /**
         * Empty title argument - Exception thrown.
         */
        @Test
        fun `Rename UserList - EmptyTitle`() {

            assertThrows<IllegalArgumentException> {
                repo.renameUserList(id, "")
            }
        }

        /**
         * Empty ID argument - Exception thrown.
         */
        @Test
        fun `Rename UserList - Empty ID`() {
            assertThrows<IllegalArgumentException> {
                repo.renameUserList("", title)
            }
        }


        /**
         * Empty title argument - Exception thrown.
         */
        @Test
        fun `Rename Item - Empty Title`() {
            assertThrows<IllegalArgumentException> {
                repo.renameItem(id, "")
            }
        }

        /**
         * Empty ID argument - Exception thrown.
         */
        @Test
        fun `Rename Item - Empty ID`() {
            assertThrows<IllegalArgumentException> {
                repo.renameItem("", title)
            }
        }
    }


    /**
     * Empty UserListId argument - Exception thrown.
     */
    @Test
    fun `Get Items - Empty UserListId`() {
        assertThrows<IllegalArgumentException> {
            repo.getItems("")
        }
    }


    @Nested
    inner class UserListPosition {
        @Test
        fun `Update UserList Position`() {
            val newPosition = 1
            val oldPosition = 5

            repo.updateUserListPosition(userList, oldPosition, newPosition)

            verify { remoteRepo.updateUserListPosition(userList, oldPosition, newPosition) }
        }

        /**
         * Both positions are the same - method returns without invoking the RemoteRepo.
         */
        @Test
        fun `Update UserList Position - Both Positions are the same`() {
            val position = 0

            repo.updateUserListPosition(userList, position, position)

            verify(exactly = 0) { remoteRepo.updateUserListPosition(userList, position, position) }
        }

        /**
         * Both positions are negative - Exception is thrown.
         */
        @Test
        fun updateUserListPosition_BothPositionsAreNegative() {
            val position1 = -1
            val position2 = -10

            assertThrows<IllegalArgumentException> {
                repo.updateUserListPosition(userList, position1, position2)
            }
        }

        /**
         * First position is negative - Exception is thrown.
         */
        @Test
        fun `Update UserList Position - First Position is negative`() {
            val position1 = -1
            val position2 = 10

            assertThrows<IllegalArgumentException> {
                repo.updateUserListPosition(userList, position1, position2)
            }

        }

        /**
         * Second position is negative - Exception is thrown.
         */
        @Test
        fun `Update UserList Position - Second Position is negative`() {
            val position1 = 1
            val position2 = -10

            assertThrows<IllegalArgumentException> {
                repo.updateUserListPosition(userList, position1, position2)
            }
        }
    }

    @Nested
    inner class ItemPosition {
        @Test
        fun `Update Item Position`() {
            val newPosition = 1
            val oldPosition = 5

            repo.updateItemPosition(item, oldPosition, newPosition)

            verify { remoteRepo.updateItemPosition(item, oldPosition, newPosition) }
        }

        /***
         * Both positions are the same - returns without invoking the RemoteRepo.
         */
        @Test
        fun `Update Item Position - Both Positions are the same`() {
            val position = 0

            repo.updateItemPosition(item, position, position)

            verify(exactly = 0) { remoteRepo.updateItemPosition(item, position, position) }
        }

        /**
         * Both positions are negative - Exception is thrown.
         */
        @Test
        fun `Update Item Position - Both Positions are negative`() {
            val position1 = -1
            val position2 = -10

            assertThrows<IllegalArgumentException> {
                repo.updateItemPosition(item, position1, position2)
            }
        }

        /**
         * First position is negative - Exception is thrown.
         */
        @Test
        fun updateItemPosition_FirstPositionIsNegative() {
            val position1 = -1
            val position2 = 10

            assertThrows<IllegalArgumentException> {
                repo.updateItemPosition(item, position1, position2)
            }
        }

        /**
         * Second position is negative - Exception is thrown.
         */
        @Test
        fun updateItemPosition_SecondPositionIsNegative() {
            val position1 = 1
            val position2 = -10

            assertThrows<IllegalArgumentException> {
                repo.updateItemPosition(item, position1, position2)
            }
        }
    }
}